test/Npgsql.Tests/Replication/PgOutputReplicationTests.cs

```diff
@@ -0,0 +1,537 @@
+CREATE PUBLICATION {publicationName} FOR TABLE {tableName};
+");
+                    var rc = await OpenReplicationConnectionAsync();
+                    var slot = await rc.CreateReplicationSlot(slotName);
+                    await c.ExecuteNonQueryAsync($"INSERT INTO {tableName} (name) VALUES ('val1'), ('val2')");
+
+                    using var streamingCts = new CancellationTokenSource();
+                    var replicationTask = Task.Run(async () =>
+                    {
+                        await foreach (var msg in (await rc.StartReplication(slot, new NpgsqlPgOutputPluginOptions(publicationName), cancellationToken: streamingCts.Token)).WithCancellation(streamingCts.Token))
+                            messages.Enqueue(msg);
+                    }, CancellationToken.None);
+
+                    // Begin Transaction
+                    var message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<BeginMessage>());
+
+                    // Relation
+                    message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<RelationMessage>());
+                    var relMsg = (RelationMessage)message!;
+                    Assert.That(relMsg.RelationReplicaIdentitySetting, Is.EqualTo('d'));
+                    Assert.That(relMsg.Namespace, Is.EqualTo("public"));
+                    Assert.That(relMsg.RelationName, Is.EqualTo(tableName));
+                    Assert.That(relMsg.Columns.Length, Is.EqualTo(2));
+                    Assert.That(relMsg.Columns[0].ColumnName, Is.EqualTo("id"));
+                    Assert.That(relMsg.Columns[1].ColumnName, Is.EqualTo("name"));
+
+                    // Insert first value
+                    message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<InsertMessage>());
+                    var insertMsg = (InsertMessage)message!;
+                    Assert.That(insertMsg.NewRow.Length, Is.EqualTo(2));
+                    Assert.That(insertMsg.NewRow[0].Value, Is.EqualTo("1"));
+                    Assert.That(insertMsg.NewRow[1].Value, Is.EqualTo("val1"));
+
+                    // Insert second value
+                    message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<InsertMessage>());
+                    insertMsg = (InsertMessage)message!;
+                    Assert.That(insertMsg.NewRow.Length, Is.EqualTo(2));
+                    Assert.That(insertMsg.NewRow[0].Value, Is.EqualTo("2"));
+                    Assert.That(insertMsg.NewRow[1].Value, Is.EqualTo("val2"));
+
+                    // Commit Transaction
+                    message = await DequeueMessage(messages);
+                    Assert.That(message, Is.TypeOf<CommitMessage>());
+
+                    streamingCts.Cancel();
+                    Assert.That(async () => await replicationTask, Throws.Exception.AssignableTo<OperationCanceledException>()
```

Comment:
This stuff is duplicated across all tests; in a test which doesn't specifically exercise cancellation/termination, we should ideally not have to do this.

We could have some sort of wrapper around StartReplication (above), which returns an IDisposable that handles all the termination at the end. That IDisposable would also have the IAsyncEnumerable on it (returned from the real StartReplication), or maybe even just the IAsyncEnumerator directly (since that's what we'd use, see comment above).

Something similar could possibly be done with tearing down the replication slot (although maybe that can be done in the wrapping SafeReplicationTest?)

Reply:
👍

Reply:
Scoped out to #3302